### PR TITLE
Refactor test-cases to use data arrays

### PR DESCRIPTION
**Pull request by michaeljones**
Hi, I realise this is quite subjective. It only took me a few minutes so I'm happy to throw it away if you don't like this kind of approach in tests. My feeling is that it is nice to have clear blocks of examples of good email addresses and bad addresses and to add more to those as needed.

I've not added any more cases here to avoid complications if you don't like the approach. Otherwise I'll look into adding more cases with either this approach or the previous one.

### CODE REVIEW

**Comment by michaeljones** [comment]
This includes your other two commits so as to avoid conflicts but I'm happy to rebase when those are merged or however you like.

**Comment by michaeljones** [comment]
Ok, I played around a bit more and changed the testing for valid emails addresses to use 'parse' and check the result. I've also added some examples of email addresses that also pass but explicitly check certain things.

The only two that I've found that fail so far are: 
- email@[123.123.123.123]
- "email"@domain.com

I feel bad for adding more commits here as it might make you feel more obliged to take all the changes together. But I've enjoyed making the changes and it hasn't taken much time so please feel free to dismiss the lot or just take what you'd like.

File: tests/EmailTest.elm [hunk 1]
```diff
@@ -18,36 +18,130 @@ suite =
...
-                \_ ->
-                    isValid "hello@"
-                        |> Expect.equal False
-            , test "missing local" <|
-                \_ ->
-                    isValid "@world.com"
-                        |> Expect.equal False
-            ]
+        , describe "valid emails" <| validEmailTests
+        , describe "invalid emails" <| invalidEmailTests
```
Comment (panthershark) [inline tests/EmailTest.elm:22]: This code organization is nice in the tests. Thanks!

File: src/Email.elm [hunk 2]
```diff
@@ -68,19 +68,29 @@ checkLocal str =
...
+        problem "local part is empty"
+
+    else if String.startsWith "." str then
+        problem "local part cannot start with ."
+
+    else if String.endsWith "." str then
+        problem "local part cannot end with ."
+
+    else if String.contains ".." str then
+        problem "local part cannot contain .."
```
Comment (panthershark) [inline src/Email.elm]: The specific error messages are really nice to end users so they can see why it failed.

**Comment by panthershark** [comment]
I really like the updates. I merged and pushed a new version. Thanks for the help!

File: tests/EmailTest.elm (hunk 1 above)
Comment (michaeljones) [reply tests/EmailTest.elm:22]: Glad you like it! Thanks for merging :)

**Pull request merged**